pandas/core/generic.py

```diff
@@ -5960,19 +5960,39 @@ def fillna(
                 )
 
             elif isinstance(value, (dict, ABCSeries)):
+                result = self if inplace else self.copy()
+
                 if axis == 1:
-                    raise NotImplementedError(
-                        "Currently only can fill "
-                        "with dict/Series column "
-                        "by column"
-                    )
+                    # To access column base
+                    result = result.T
 
-                result = self if inplace else self.copy()
+                columns = result.columns
                 for k, v in value.items():
```

Comment:
Can you just enumerate the columns and fill? Would be a lot simpler than trying to deduce later if access by label results in a 1D or 2D object